pkg/cloudprovider/provider/azure/provider.go

```diff
@@ -0,0 +1,581 @@
+	if err != nil {
+		return nil, fmt.Errorf("failed to create interfaces client: %v", err)
+	}
+
+	netIf, err := ifClient.Get(ctx, c.ResourceGroup, ifaceName, "")
+	if err != nil {
+		return nil, fmt.Errorf("failed to get interface %q: %s", ifaceName, err.Error())
+	}
+
+	ipAddresses := make([]string, 0)
+
+	if netIf.IPConfigurations != nil {
+		for _, conf := range *netIf.IPConfigurations {
+			var name string
+			if conf.Name != nil {
+				name = *conf.Name
+			} else {
+				glog.Warning("IP configuration of NIC %q was returned with no name, trying to dissect the ID.", ifaceName)
+				if conf.ID == nil {
+					return nil, fmt.Errorf("IP configuration of NIC %q was returned with no ID", ifaceName)
+				}
+				splitConfID := strings.Split(*conf.ID, "/")
+				name = splitConfID[len(splitConfID)-1]
+			}
+
+			addrStrings, err := getIPAddressStrings(ctx, c, name)
+			if err != nil {
+				return nil, fmt.Errorf("failed to retrieve IP string for IP %q: %s", name, err)
+			}
+
+			ipAddresses = append(ipAddresses, addrStrings...)
+		}
+	}
+
+	return ipAddresses, nil
+}
+
+func getIPAddressStrings(ctx context.Context, c *config, addrName string) ([]string, error) {
+	ipClient, err := getIPClient(c)
+	if err != nil {
+		return nil, fmt.Errorf("failed to create IP address client: %v", err)
+	}
+
+	ip, err := ipClient.Get(ctx, c.ResourceGroup, addrName, "")
+	if err != nil {
+		return nil, fmt.Errorf("failed to get IP %q: %s", addrName, err)
+	}
+
+	if ip.IPConfiguration == nil {
+		return nil, fmt.Errorf("IP %q has nil IPConfiguration", addrName)
+	}
+
+	ipAddresses := make([]string, 0)
+	if ip.IPConfiguration.PublicIPAddress != nil && ip.IPConfiguration.PublicIPAddress.IPAddress != nil {
+		ipAddresses = append(ipAddresses, *ip.IPConfiguration.PublicIPAddress.IPAddress)
+	}
+	if ip.IPConfiguration.PrivateIPAddress != nil {
+		ipAddresses = append(ipAddresses, *ip.IPConfiguration.PrivateIPAddress)
+	}
+
+	return ipAddresses, nil
+}
+
+func (p *provider) AddDefaults(spec v1alpha1.MachineSpec) (v1alpha1.MachineSpec, bool, error) {
+	return spec, false, nil
+}
+
+func (p *provider) Create(machine *v1alpha1.Machine, userdata string) (instance.Instance, error) {
+	config, providerCfg, err := p.getConfig(machine.Spec.ProviderConfig)
+	if err != nil {
+		return nil, cloudprovidererrors.TerminalError{
+			Reason:  v1alpha1.InvalidConfigurationMachineError,
+			Message: fmt.Sprintf("Failed to parse MachineSpec, due to %v", err),
+		}
+	}
+
+	vmClient, err := getVMClient(config)
+	if err != nil {
+		return nil, fmt.Errorf("failed to create VM client: %v", err)
+	}
+
+	osRef, err := getOSImageReference(providerCfg.OperatingSystem)
+	if err != nil {
+		return nil, err
+	}
+
+	// We genete a random SSH key, since Azure won't let us create a VM without an SSH key or a password
+	key, err := ssh.NewSSHKey()
+	if err != nil {
+		return nil, fmt.Errorf("failed to generate ssh key: %v", err)
+	}
+
+	ifaceName := machine.Spec.Name + "-netiface"
```

Comment:
Ugh, Azure doesnt allow to implicitly create the interface when creating the instance?

Reply:
Not that I've seen.